# San Francisco Crime Project

* Author: Kevin Chuang [@k-chuang](https://www.github.com/k-chuang)
* Created on: August 25, 2018
* Description: Data analysis, exploration, visualization, and data mining on crime in SF
* Original dataset: [SF Gov Crime dataset](https://data.sfgov.org/Public-Safety/-Change-Notice-Police-Department-Incidents/tmnf-yvry/about)
* Kaggle dataset: [Kaggle SF Crime](https://www.kaggle.com/c/sf-crime/data)

---------------

# Table of Contents

- Introduction
    - SF Crime Dataset
- Basic Preparation
    - Import libraries
    - Load data
- Data Exploration/Analysis Extension
- Data Preprocessing
    - Data Imputation/Removal
    - Feature Engineering
    - Feature Encoding
- Build Machine Learning Models
    - Train different baseline models
    - Analyze results
- Model Selection
- Hyperparameter tuning
- Train Model with optimal hyperparameters
- Feature Selection
    - Feature Importance
    - Feature Removal
- Train Final Model
- Model Evaluation
- Summary
- Kaggle Submission
- Conclusion

# Introduction


## SF Crime Dataset

This dataset contains incidents derived from SFPD Crime Incident Reporting system. The data ranges from 1/1/2003 to 5/13/2015. The training set and test set rotate every week, meaning week 1,3,5,7... belong to test set, week 2,4,6,8 belong to training set. The goal is to try to predict the category of crime that occurred in the city of San Francisco. 

### Data Fields
- **Dates** - timestamp of the crime incident
- **Category** - category of the crime incident (only in train.csv). This is the target variable you are going to predict.
- **Descript** - detailed description of the crime incident (only in train.csv)
- **DayOfWeek** - the day of the week
- **PdDistrict** - name of the Police Department District
- **Resolution** - how the crime incident was resolved (only in train.csv)
- **Address** - the approximate street address of the crime incident 
- **X** - Longitude
- **Y** - Latitude


In this juypter notebook, I will go through the whole process, end-to-end, of creating a machine learning model on the open source San Francisco Crime dataset. This includes data exploration & analysis, data preprocessing (huge part of this project and includes feature engineering), trying out different ML algorithms and determining the optimal ML model, tuning the hyperparameters of that model, and finally, evaluating the chosen model in terms of multiclass log loss. 

Since this is an old Kaggle competition, I will refrain from looking online for resources or old Kaggle kernels. The plan is to get better at coding an end to end data science project and to familiarize myself with the Python data science libraries. Also, I hope to learn some interesting things and discover some cool patterns or ideas using this dataset. Well, here goes nothing!



## Import libraries

In [1]:
__author__ = 'Kevin Chuang (https://www.github.com/k-chuang)' 

# linear algebra
import numpy as np 

# data processing
import pandas as pd 

# data visualization
import seaborn as sns
%matplotlib inline
from matplotlib import pyplot as plt
from matplotlib import style

# Algorithms
from sklearn import linear_model
from sklearn.linear_model import SGDClassifier, LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import Perceptron
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC, LinearSVC
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.ensemble import AdaBoostClassifier
from sklearn.neural_network import MLPClassifier
from xgboost import XGBClassifier
import xgboost as xgb

# Preprocessing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler, OneHotEncoder

# Metrics 
from sklearn.metrics import log_loss
from sklearn.model_selection import cross_val_score

# Model Selection & Hyperparameter tuning
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV, StratifiedKFold
from skopt import BayesSearchCV
from skopt.space  import Real, Categorical, Integer


# Clustering
from sklearn.cluster import KMeans

# Mathematical Functions
import math
import re

## Load data

In [2]:
train_df = pd.read_csv('data/train.csv')
test_df = pd.read_csv('data/test.csv')

In [3]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 9 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   Dates       878049 non-null  object 
 1   Category    878049 non-null  object 
 2   Descript    878049 non-null  object 
 3   DayOfWeek   878049 non-null  object 
 4   PdDistrict  878049 non-null  object 
 5   Resolution  878049 non-null  object 
 6   Address     878049 non-null  object 
 7   X           878049 non-null  float64
 8   Y           878049 non-null  float64
dtypes: float64(2), object(7)
memory usage: 60.3+ MB


# Data Exploration & Analysis Extension

- Complete data exploration & visualizations are located in jupyter notebook: [kaggle-sf-crime-data-exploration.ipynb](kaggle-sf-crime-data-exploration.ipynb)
- This dataset suffers from **imbalanced classes** (**TREA** has 6 occurrences while **LARCENY/THEFT** has 1,749,000 occurrences)
    - There are a couple ways to deal with imbalanced classes, such as:
        - Changing performance metric (Do not use accuracy, use a confusion matrix, precision, recall, F1 score, ROC curves)
        - Resample dataset (Oversample under-represented classes, and undersample over-represented classes)
        - Try different ML algorithms that can handle imbalanced classes
            - Decision Trees (Random Forests/XGBoost) often perform well on imbalanced classes (due to splitting rules)

In [4]:
train_df.head(8)

Dates        Category                        Descript  \
0  2015-05-13 23:53:00        WARRANTS                  WARRANT ARREST   
1  2015-05-13 23:53:00  OTHER OFFENSES        TRAFFIC VIOLATION ARREST   
2  2015-05-13 23:33:00  OTHER OFFENSES        TRAFFIC VIOLATION ARREST   
3  2015-05-13 23:30:00   LARCENY/THEFT    GRAND THEFT FROM LOCKED AUTO   
4  2015-05-13 23:30:00   LARCENY/THEFT    GRAND THEFT FROM LOCKED AUTO   
5  2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM UNLOCKED AUTO   
6  2015-05-13 23:30:00   VEHICLE THEFT               STOLEN AUTOMOBILE   
7  2015-05-13 23:30:00   VEHICLE THEFT               STOLEN AUTOMOBILE   

   DayOfWeek PdDistrict      Resolution                    Address  \
0  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
1  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
2  Wednesday   NORTHERN  ARREST, BOOKED  VANNESS AV / GREENWICH ST   
3  Wednesday   NORTHERN            NONE   1500 Block of LOMBARD ST   
4  Wednesday       PARK            NONE  100 Block of BRODERICK ST   
5  Wednesday  INGLESIDE            NONE        0 Block of TEDDY AV   
6  Wednesday  INGLESIDE            NONE        AVALON AV / PERU AV   
7  Wednesday    BAYVIEW            NONE   KIRKWOOD AV / DONAHUE ST   

            X          Y  
0 -122.425892  37.774599  
1 -122.425892  37.774599  
2 -122.424363  37.800414  
3 -122.426995  37.800873  
4 -122.438738  37.771541  
5 -122.403252  37.713431  
6 -122.423327  37.725138  
7 -122.371274  37.727564

In [5]:
train_df.columns.values

array(['Dates', 'Category', 'Descript', 'DayOfWeek', 'PdDistrict',
       'Resolution', 'Address', 'X', 'Y'], dtype=object)

In [6]:
# set show nulls to True
train_df.info(verbose=True, null_counts=True)

/tmp/ipykernel_702747/1066574219.py:2: FutureWarning: null_counts is deprecated. Use show_counts instead
  train_df.info(verbose=True, null_counts=True)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 9 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   Dates       878049 non-null  object 
 1   Category    878049 non-null  object 
 2   Descript    878049 non-null  object 
 3   DayOfWeek   878049 non-null  object 
 4   PdDistrict  878049 non-null  object 
 5   Resolution  878049 non-null  object 
 6   Address     878049 non-null  object 
 7   X           878049 non-null  float64
 8   Y           878049 non-null  float64
dtypes: float64(2), object(7)
memory usage: 60.3+ MB


------------
### Things we learned thus far:

- 878,049 instances in training set (or recorded crime instances in SF)
- 9 columns (8 potential features + 1 label (Category))
- Data types:
    - 2 columns with float values
    - 7 objects
- There are no null (NaN) values (Yay!)

In [7]:
## Count number of observations for each crime 
train_df['Category'].value_counts()

LARCENY/THEFT                  174900
OTHER OFFENSES                 126182
NON-CRIMINAL                    92304
ASSAULT                         76876
DRUG/NARCOTIC                   53971
VEHICLE THEFT                   53781
VANDALISM                       44725
WARRANTS                        42214
BURGLARY                        36755
SUSPICIOUS OCC                  31414
MISSING PERSON                  25989
ROBBERY                         23000
FRAUD                           16679
FORGERY/COUNTERFEITING          10609
SECONDARY CODES                  9985
WEAPON LAWS                      8555
PROSTITUTION                     7484
TRESPASS                         7326
STOLEN PROPERTY                  4540
SEX OFFENSES FORCIBLE            4388
DISORDERLY CONDUCT               4320
DRUNKENNESS                      4280
RECOVERED VEHICLE                3138
KIDNAPPING                       2341
DRIVING UNDER THE INFLUENCE      2268
RUNAWAY                          1946
LIQUOR LAWS 

In [8]:
## Count number of observations of crime for each PD District
train_df['PdDistrict'].value_counts()

SOUTHERN      157182
MISSION       119908
NORTHERN      105296
BAYVIEW        89431
CENTRAL        85460
TENDERLOIN     81809
INGLESIDE      78845
TARAVAL        65596
PARK           49313
RICHMOND       45209
Name: PdDistrict, dtype: int64

In [9]:
## Count number of observations for each day of week
train_df['DayOfWeek'].value_counts()

Friday       133734
Wednesday    129211
Saturday     126810
Thursday     125038
Tuesday      124965
Monday       121584
Sunday       116707
Name: DayOfWeek, dtype: int64

In [10]:
## Count number of observations for Resolution feature
train_df['Resolution'].value_counts()

NONE                                      526790
ARREST, BOOKED                            206403
ARREST, CITED                              77004
LOCATED                                    17101
PSYCHOPATHIC CASE                          14534
UNFOUNDED                                   9585
JUVENILE BOOKED                             5564
COMPLAINANT REFUSES TO PROSECUTE            3976
DISTRICT ATTORNEY REFUSES TO PROSECUTE      3934
NOT PROSECUTED                              3714
JUVENILE CITED                              3332
PROSECUTED BY OUTSIDE AGENCY                2504
EXCEPTIONAL CLEARANCE                       1530
JUVENILE ADMONISHED                         1455
JUVENILE DIVERTED                            355
CLEARED-CONTACT JUVENILE FOR MORE INFO       217
PROSECUTED FOR LESSER OFFENSE                 51
Name: Resolution, dtype: int64

In [11]:
train_df[['X','Y']].describe()

X              Y
count  878049.000000  878049.000000
mean     -122.422616      37.771020
std         0.030354       0.456893
min      -122.513642      37.707879
25%      -122.432952      37.752427
50%      -122.416420      37.775421
75%      -122.406959      37.784369
max      -120.500000      90.000000

**There seems to be an invalid coordinates (max) 90 (latitude) or -120.5 (longitude) does not seem to be a valid coordinate in San Francisco. We must fix these values for this feature.**

# Data Preprocessing

- Data cleaning
    - imputation or removal of outlier values
- Feature Engineering (Feature Creation)
- Feature Encoding
    - **Integer encode** or **label encode** ordinal categorical features that maintain order (Year, Business Quarter, Block/Street Number)
    - Usually: 
        - **One hot encode** nominal categorical features (DayOfWeek, PdDistrict, StreetType, Category)
            - mainly for logistic regression
        - However, Random Forests & Boosting algorithms can handle nominal categorical features directly, so we just **integer encode** these features.

## Data Cleaning

- Data removal
- Data imputation

In [12]:
train_df[train_df['Y'] == train_df['Y'].max()]

Dates               Category  \
660485  2005-12-30 17:00:00          LARCENY/THEFT   
660711  2005-12-30 00:34:00                ASSAULT   
660712  2005-12-30 00:34:00                ASSAULT   
661106  2005-12-29 00:07:00           NON-CRIMINAL   
666430  2005-11-30 11:25:00         OTHER OFFENSES   
...                     ...                    ...   
844995  2003-06-11 08:49:00         OTHER OFFENSES   
845842  2003-06-09 09:25:00         OTHER OFFENSES   
852880  2003-05-02 01:00:00  SEX OFFENSES FORCIBLE   
857248  2003-04-14 16:30:00                ROBBERY   
871198  2003-02-05 12:00:00          LARCENY/THEFT   

                                     Descript  DayOfWeek  PdDistrict  \
660485           GRAND THEFT FROM LOCKED AUTO     Friday  TENDERLOIN   
660711            INFLICT INJURY ON COHABITEE     Friday     BAYVIEW   
660712   AGGRAVATED ASSAULT WITH BODILY FORCE     Friday     BAYVIEW   
661106           AIDED CASE, MENTAL DISTURBED   Thursday  TENDERLOIN   
666430                      TRAFFIC VIOLATION  Wednesday  TENDERLOIN   
...                                       ...        ...         ...   
844995  DRIVERS LICENSE, SUSPENDED OR REVOKED  Wednesday   INGLESIDE   
845842  DRIVERS LICENSE, SUSPENDED OR REVOKED     Monday   INGLESIDE   
852880            FORCIBLE RAPE, BODILY FORCE     Friday    SOUTHERN   
857248       ROBBERY ON THE STREET, STRONGARM     Monday     BAYVIEW   
871198           PETTY THEFT FROM LOCKED AUTO  Wednesday    SOUTHERN   

                              Resolution  \
660485                              NONE   
660711                    ARREST, BOOKED   
660712                    ARREST, BOOKED   
661106                 PSYCHOPATHIC CASE   
666430                     ARREST, CITED   
...                                  ...   
844995                     ARREST, CITED   
845842                     ARREST, CITED   
852880  COMPLAINANT REFUSES TO PROSECUTE   
857248  COMPLAINANT REFUSES TO PROSECUTE   
871198                              NONE   

                                        Address      X     Y  
660485              5THSTNORTH ST / OFARRELL ST -120.5  90.0  
660711          JAMESLICKFREEWAY HY / SILVER AV -120.5  90.0  
660712          JAMESLICKFREEWAY HY / SILVER AV -120.5  90.0  
661106                  5THSTNORTH ST / EDDY ST -120.5  90.0  
666430                 5THSTNORTH ST / ELLIS ST -120.5  90.0  
...                                         ...    ...   ...  
844995  JAMES LICK FREEWAY HY / CESAR CHAVEZ ST -120.5  90.0  
845842  JAMES LICK FREEWAY HY / CESAR CHAVEZ ST -120.5  90.0  
852880           3RD ST / JAMES LICK FREEWAY HY -120.5  90.0  
857248                     GILMAN AV / FITCH ST -120.5  90.0  
871198      SPEAR ST / THE EMBARCADERO SOUTH ST -120.5  90.0  

[67 rows x 9 columns]

I notice that there are 108 rows with incorrect coordinates, and they seem to be the exact same two coordinates (90, -120.5). There are many ways to handle this. We need to do data imputation, which can be done several ways. For now, I will randomly sample from a normal distribution with the range of a standard deviation from the mean. However, I could use a linear regression model to predict the latitude and longitude values (based on other variables such as PD district?) and use that to impute the bad / inconsistent data points.

Another method is to completely remove this data. Since I already have a lot of data, and I do not want this incorrect data to affect my results, I could remove them. However, I will stick with data imputation.

In [13]:
train_df['Y'].replace(to_replace= train_df['Y'].max() ,value=np.nan, inplace=True)
train_df['X'].replace(to_replace= train_df['X'].max() ,value=np.nan, inplace=True)
test_df['Y'].replace(to_replace= test_df['Y'].max() ,value=np.nan, inplace=True)
test_df['X'].replace(to_replace= test_df['X'].max() ,value=np.nan, inplace=True)

In [14]:
train_df.isnull().sum()

Dates          0
Category       0
Descript       0
DayOfWeek      0
PdDistrict     0
Resolution     0
Address        0
X             67
Y             67
dtype: int64

In [15]:
test_df.isnull().sum()

Id             0
Dates          0
DayOfWeek      0
PdDistrict     0
Address        0
X             76
Y             76
dtype: int64

In [16]:
data = [train_df, test_df]

for dataset in data:
    mean_X = dataset["X"].mean()
    std_X = dataset["X"].std()
    mean_Y = dataset["Y"].mean()
    std_Y = dataset["Y"].std()
    max_X = mean_X + std_X
    min_X = mean_X - std_X
    max_Y = mean_Y + std_Y
    min_Y = mean_Y - std_Y

    # Both X and Y will have the same null so just use Y
    is_null = dataset['Y'].isnull().sum()
    random_X = (max_X - min_X) * np.random.randn(is_null) + min_X
    random_Y = (max_Y - min_Y) * np.random.randn(is_null) + min_Y

    X_slice = dataset['X'].copy()
    Y_slice = dataset['Y'].copy()
    X_slice[np.isnan(X_slice)] = random_X
    Y_slice[np.isnan(Y_slice)] = random_Y
    dataset['X'] = X_slice
    dataset['Y'] = Y_slice


In [17]:
train_df[['X', 'Y']].describe()

X              Y
count  878049.000000  878049.000000
mean     -122.422765      37.767032
std         0.025288       0.024169
min      -122.545017      37.621260
25%      -122.432952      37.752427
50%      -122.416422      37.775421
75%      -122.406959      37.784368
max      -122.332641      37.864901

In [18]:
len(train_df)

878049

In [19]:
test_df[['X', 'Y']].describe()

X              Y
count  884262.000000  884262.000000
mean     -122.422860      37.766985
std         0.025350       0.024183
min      -122.578772      37.636054
25%      -122.433075      37.752357
50%      -122.416517      37.775421
75%      -122.406959      37.784353
max      -122.311837      37.833406

In [20]:
len(test_df)

884262

# Feature Engineering

- Let's create some new features from the data that exists in the current feature space
- There are a couple categories of features:
    - Temporal features
    - Spatial features

## Temporal Features
We want to have a column for Time, so we must parse through the 'Dates' feature to create the 'Time' feature


In [21]:
# Transform the Date into a python datetime object.
train_df["Dates"] = pd.to_datetime(train_df["Dates"], format="%Y-%m-%d %H:%M:%S")
test_df["Dates"] = pd.to_datetime(test_df["Dates"], format="%Y-%m-%d %H:%M:%S")

In [22]:
# Minute
train_df["Minute"] = train_df["Dates"].map(lambda x: x.minute)
test_df["Minute"] = test_df["Dates"].map(lambda x: x.minute)

In [23]:
# Hour
train_df["Hour"] = train_df["Dates"].map(lambda x: x.hour)
test_df["Hour"] = test_df["Dates"].map(lambda x: x.hour)

In [24]:
# Day
train_df["Day"] = train_df["Dates"].map(lambda x: x.day)
test_df["Day"] = test_df["Dates"].map(lambda x: x.day)

In [25]:
# Month
train_df["Month"] = train_df["Dates"].map(lambda x: x.month)
test_df["Month"] = test_df["Dates"].map(lambda x: x.month)

In [26]:
# Year
train_df["Year"] = train_df["Dates"].map(lambda x: x.year)
test_df["Year"] = test_df["Dates"].map(lambda x: x.year)

In [27]:
# Hour Zone 0 - Pass midnight, 1 - morning, 2 - afternoon, 3 - dinner / sun set, 4 - night
def get_hour_zone(hour):
    if hour >= 2 and hour < 8: 
        return 0
    elif hour >= 8 and hour < 12: 
        return 1
    elif hour >= 12 and hour < 18: 
        return 2
    elif hour >= 18 and hour < 22: 
        return 3
    elif hour < 2 or hour >= 22: 
        return 4
    
train_df["Hour_Zone"] = train_df["Hour"].map(get_hour_zone)
test_df["Hour_Zone"] = test_df["Hour"].map(get_hour_zone)

In [28]:
# Add Week of Year
train_df["WeekOfYear"] = train_df["Dates"].map(lambda x: int(x.weekofyear / 2) - 1)
test_df["WeekOfYear"] = test_df["Dates"].map(lambda x: int(x.weekofyear / 2))

print(sorted(train_df['WeekOfYear'].unique()))
print(sorted(test_df['WeekOfYear'].unique()))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]


In [29]:
train_df.head()

Dates        Category                      Descript  \
0 2015-05-13 23:53:00        WARRANTS                WARRANT ARREST   
1 2015-05-13 23:53:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
2 2015-05-13 23:33:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
3 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   
4 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   

   DayOfWeek PdDistrict      Resolution                    Address  \
0  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
1  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
2  Wednesday   NORTHERN  ARREST, BOOKED  VANNESS AV / GREENWICH ST   
3  Wednesday   NORTHERN            NONE   1500 Block of LOMBARD ST   
4  Wednesday       PARK            NONE  100 Block of BRODERICK ST   

            X          Y  Minute  Hour  Day  Month  Year  Hour_Zone  \
0 -122.425892  37.774599      53    23   13      5  2015          4   
1 -122.425892  37.774599      53    23   13      5  2015          4   
2 -122.424363  37.800414      33    23   13      5  2015          4   
3 -122.426995  37.800873      30    23   13      5  2015          4   
4 -122.438738  37.771541      30    23   13      5  2015          4   

   WeekOfYear  
0           9  
1           9  
2           9  
3           9  
4           9

### Holiday Feature

- Certain crimes may be more apparent on holidays

In [30]:
from pandas.tseries.holiday import USFederalHolidayCalendar as calendar

# Training set
cal = calendar()
holidays = cal.holidays(start=train_df['Dates'].min(), end=train_df['Dates'].max())
train_df['Holiday'] = train_df['Dates'].dt.date.astype('datetime64').isin(holidays)

In [31]:
# Test set
cal = calendar()
holidays = cal.holidays(start=test_df['Dates'].min(), end=test_df['Dates'].max())
test_df['Holiday'] = test_df['Dates'].dt.date.astype('datetime64').isin(holidays)

In [32]:
len(train_df[train_df['Holiday'] == True])

25653

In [33]:
len(test_df[test_df['Holiday'] == True])

18316

### Business Hours Feature

- There should be an effect of business hours on the type of crime committed
- Let's create a binary feature where:
    - 1 is typical business hours [8:00AM - 6:00PM]
    - 0 is not business hours [6:01PM - 7:59 AM]

In [34]:
from datetime import datetime, time

def time_in_range(start, end, x):
    """Return true if x is in the inclusive range [start, end]"""
    if start <= end:
        return start <= x <= end
    else:
        return start <= x or x <= end

def map_business_hours(date):
    
    # Convert military time to AM & PM
    time_parsed = date.time()
    business_start = time(8, 0, 0)
    business_end = time(18, 0, 0)
    
    if time_in_range(business_start, business_end, time_parsed):
        return 1
    else:
        return 0
    
train_df['BusinessHour'] = train_df['Dates'].map(map_business_hours).astype('uint8')
test_df['BusinessHour'] = test_df['Dates'].map(map_business_hours).astype('uint8')

In [35]:
train_df['BusinessHour'].value_counts()

1    455215
0    422834
Name: BusinessHour, dtype: int64

In [36]:
train_df.head(10)

Dates        Category                        Descript  \
0 2015-05-13 23:53:00        WARRANTS                  WARRANT ARREST   
1 2015-05-13 23:53:00  OTHER OFFENSES        TRAFFIC VIOLATION ARREST   
2 2015-05-13 23:33:00  OTHER OFFENSES        TRAFFIC VIOLATION ARREST   
3 2015-05-13 23:30:00   LARCENY/THEFT    GRAND THEFT FROM LOCKED AUTO   
4 2015-05-13 23:30:00   LARCENY/THEFT    GRAND THEFT FROM LOCKED AUTO   
5 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM UNLOCKED AUTO   
6 2015-05-13 23:30:00   VEHICLE THEFT               STOLEN AUTOMOBILE   
7 2015-05-13 23:30:00   VEHICLE THEFT               STOLEN AUTOMOBILE   
8 2015-05-13 23:00:00   LARCENY/THEFT    GRAND THEFT FROM LOCKED AUTO   
9 2015-05-13 23:00:00   LARCENY/THEFT    GRAND THEFT FROM LOCKED AUTO   

   DayOfWeek PdDistrict      Resolution                        Address  \
0  Wednesday   NORTHERN  ARREST, BOOKED             OAK ST / LAGUNA ST   
1  Wednesday   NORTHERN  ARREST, BOOKED             OAK ST / LAGUNA ST   
2  Wednesday   NORTHERN  ARREST, BOOKED      VANNESS AV / GREENWICH ST   
3  Wednesday   NORTHERN            NONE       1500 Block of LOMBARD ST   
4  Wednesday       PARK            NONE      100 Block of BRODERICK ST   
5  Wednesday  INGLESIDE            NONE            0 Block of TEDDY AV   
6  Wednesday  INGLESIDE            NONE            AVALON AV / PERU AV   
7  Wednesday    BAYVIEW            NONE       KIRKWOOD AV / DONAHUE ST   
8  Wednesday   RICHMOND            NONE           600 Block of 47TH AV   
9  Wednesday    CENTRAL            NONE  JEFFERSON ST / LEAVENWORTH ST   

            X          Y  Minute  Hour  Day  Month  Year  Hour_Zone  \
0 -122.425892  37.774599      53    23   13      5  2015          4   
1 -122.425892  37.774599      53    23   13      5  2015          4   
2 -122.424363  37.800414      33    23   13      5  2015          4   
3 -122.426995  37.800873      30    23   13      5  2015          4   
4 -122.438738  37.771541      30    23   13      5  2015          4   
5 -122.403252  37.713431      30    23   13      5  2015          4   
6 -122.423327  37.725138      30    23   13      5  2015          4   
7 -122.371274  37.727564      30    23   13      5  2015          4   
8 -122.508194  37.776601       0    23   13      5  2015          4   
9 -122.419088  37.807802       0    23   13      5  2015          4   

   WeekOfYear  Holiday  BusinessHour  
0           9    False             0  
1           9    False             0  
2           9    False             0  
3           9    False             0  
4           9    False             0  
5           9    False             0  
6           9    False             0  
7           9    False             0  
8           9    False             0  
9           9    False             0

### Season

The season feature may affect what type of crimes are commited. 
- 1 = Winter, 2 = Spring, 3 = Summer, 4 = Fall

In [37]:
train_df['Season']=(train_df['Month']%12 + 3)//3
test_df['Season']=(test_df['Month']%12 + 3)//3

In [38]:
train_df.head()

Dates        Category                      Descript  \
0 2015-05-13 23:53:00        WARRANTS                WARRANT ARREST   
1 2015-05-13 23:53:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
2 2015-05-13 23:33:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
3 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   
4 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   

   DayOfWeek PdDistrict      Resolution                    Address  \
0  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
1  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
2  Wednesday   NORTHERN  ARREST, BOOKED  VANNESS AV / GREENWICH ST   
3  Wednesday   NORTHERN            NONE   1500 Block of LOMBARD ST   
4  Wednesday       PARK            NONE  100 Block of BRODERICK ST   

            X          Y  Minute  Hour  Day  Month  Year  Hour_Zone  \
0 -122.425892  37.774599      53    23   13      5  2015          4   
1 -122.425892  37.774599      53    23   13      5  2015          4   
2 -122.424363  37.800414      33    23   13      5  2015          4   
3 -122.426995  37.800873      30    23   13      5  2015          4   
4 -122.438738  37.771541      30    23   13      5  2015          4   

   WeekOfYear  Holiday  BusinessHour  Season  
0           9    False             0       2  
1           9    False             0       2  
2           9    False             0       2  
3           9    False             0       2  
4           9    False             0       2

### Weekend

- Weekends may have effect on what types of crimes are commmited
- Weekday = 0, Weekend =1

In [39]:
# Weekend Feature

# Weekday = 0, Weekend = 1
days = {'Monday':0 ,'Tuesday':0 ,'Wednesday':0 ,'Thursday':0 ,'Friday':0, 'Saturday':1 ,'Sunday':1}

train_df['Weekend'] = train_df['DayOfWeek'].replace(days).astype('uint8')
test_df['Weekend'] = test_df['DayOfWeek'].replace(days).astype('uint8')

## Spatial Features

### Street Type

The street type can have an effect on what type of crime is committed, so we want to extract the street type from the 'Address' feature.

We have avenues, streets, ways, boulevards, highways, courts, walks, plazas, and differet number of intersections of roads/streets (Addresses with /).

In [40]:
train_df['Address'].value_counts().index

Index(['800 Block of BRYANT ST', '800 Block of MARKET ST',
       '2000 Block of MISSION ST', '1000 Block of POTRERO AV',
       '900 Block of MARKET ST', '0 Block of TURK ST', '0 Block of 6TH ST',
       '300 Block of ELLIS ST', '400 Block of ELLIS ST',
       '16TH ST / MISSION ST',
       ...
       '5TH AV / CALIFORNIA ST', 'HAZELWOOD AV / YERBABUENA AV',
       '0 Block of TROY AL', '900 Block of MARTIN LUTHER KING JR DR',
       'CLIPPER ST / PORTOLA DR', 'PRECITA AV / CESAR CHAVEZ ST',
       '0 Block of COLUSA PL', '2800 Block of KEITH ST',
       'CABRILLO ST / ARGUELLO BL', 'OFARRELL ST / CYRIL MAGNIN ST'],
      dtype='object', length=23228)

In [41]:
def find_streets(address):
    street_types = ['AV', 'ST', 'CT', 'PZ', 'LN', 'DR', 'PL', 'HY', 
                    'FY', 'WY', 'TR', 'RD', 'BL', 'WAY', 'CR', 'AL', 'I-80',  
                    'RW', 'WK','EL CAMINO DEL MAR']
    street_pattern = '|'.join(street_types)
    streets = re.findall(street_pattern, address)
    if len(streets) == 0:
        # Debug
#         print(address)
        return 'OTHER'
    elif len(streets) == 1:
        return streets[0]
    else:
#         print(address)
        return 'INT'

train_df['StreetType'] = train_df['Address'].map(find_streets)
test_df['StreetType'] = test_df['Address'].map(find_streets)


In [42]:
train_df['StreetType'].value_counts()

INT                  389995
ST                   358797
AV                    92467
BL                    13074
DR                     8200
WY                     4063
RD                     2384
PZ                     2347
CT                     2059
LN                     1356
PL                      863
HY                      819
TR                      766
I-80                    322
CR                      291
AL                      150
WAY                      55
EL CAMINO DEL MAR        21
OTHER                    14
WK                        5
RW                        1
Name: StreetType, dtype: int64

In [43]:
# Check for null values
train_df['StreetType'].isnull().sum()

0

In [44]:
train_df.head(8)

Dates        Category                        Descript  \
0 2015-05-13 23:53:00        WARRANTS                  WARRANT ARREST   
1 2015-05-13 23:53:00  OTHER OFFENSES        TRAFFIC VIOLATION ARREST   
2 2015-05-13 23:33:00  OTHER OFFENSES        TRAFFIC VIOLATION ARREST   
3 2015-05-13 23:30:00   LARCENY/THEFT    GRAND THEFT FROM LOCKED AUTO   
4 2015-05-13 23:30:00   LARCENY/THEFT    GRAND THEFT FROM LOCKED AUTO   
5 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM UNLOCKED AUTO   
6 2015-05-13 23:30:00   VEHICLE THEFT               STOLEN AUTOMOBILE   
7 2015-05-13 23:30:00   VEHICLE THEFT               STOLEN AUTOMOBILE   

   DayOfWeek PdDistrict      Resolution                    Address  \
0  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
1  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
2  Wednesday   NORTHERN  ARREST, BOOKED  VANNESS AV / GREENWICH ST   
3  Wednesday   NORTHERN            NONE   1500 Block of LOMBARD ST   
4  Wednesday       PARK            NONE  100 Block of BRODERICK ST   
5  Wednesday  INGLESIDE            NONE        0 Block of TEDDY AV   
6  Wednesday  INGLESIDE            NONE        AVALON AV / PERU AV   
7  Wednesday    BAYVIEW            NONE   KIRKWOOD AV / DONAHUE ST   

            X          Y  Minute  ...  Day  Month  Year  Hour_Zone  \
0 -122.425892  37.774599      53  ...   13      5  2015          4   
1 -122.425892  37.774599      53  ...   13      5  2015          4   
2 -122.424363  37.800414      33  ...   13      5  2015          4   
3 -122.426995  37.800873      30  ...   13      5  2015          4   
4 -122.438738  37.771541      30  ...   13      5  2015          4   
5 -122.403252  37.713431      30  ...   13      5  2015          4   
6 -122.423327  37.725138      30  ...   13      5  2015          4   
7 -122.371274  37.727564      30  ...   13      5  2015          4   

   WeekOfYear  Holiday  BusinessHour  Season  Weekend  StreetType  
0           9    False             0       2        0         INT  
1           9    False             0       2        0         INT  
2           9    False             0       2        0         INT  
3           9    False             0       2        0         INT  
4           9    False             0       2        0          ST  
5           9    False             0       2        0          AV  
6           9    False             0       2        0         INT  
7           9    False             0       2        0         INT  

[8 rows x 21 columns]

## Block Number Feature

- Let's explore the block number from address
- Block number has ordinal data type (order matters), and has spatial significance
- It seems all the block numbers are in intervals of 100
- How to categorize
    - Addresses that do not have a block number will be categorized as 0
    - Addresses with block number will be divided by 100, and added by 1 for mapping (0 is saved for addresses with no block number)
- 85 unique block numbers (including 1 where there is no block number)

In [45]:
def find_block_number(address):
    block_num_pattern = '[0-9]+\s[Block]'
    block_num = re.search(block_num_pattern, address)
    if block_num:
#         print(address)
        num_pattern = '[0-9]+'
        block_no_pos = re.search(num_pattern, address)
        # Get integer of found regular expression
        block_no = int(block_no_pos.group())
        # Convert block number by dividing by 100 and adding 1 (0 = addresses with no block)
        block_map = (block_no // 100) + 1
#         print(block_map)
        return block_map
    else:
#         print(address)
        # 
        return 0


train_df['BlockNo'] = train_df['Address'].map(find_block_number)
test_df['BlockNo'] = test_df['Address'].map(find_block_number)

In [46]:
train_df['BlockNo'].value_counts()

0     260818
1      76325
2      51917
9      51718
3      38407
       ...  
82         7
79         5
81         4
84         4
80         3
Name: BlockNo, Length: 85, dtype: int64

## X, Y Coordinates

- Normalize and scale the X and Y coordinates
- I use **K-Means clustering** to create a new feature for the longitude and latitude by grouping clusters of points based on Euclidean distances.
- X = longitude, Y = latitude
- I also extract more spatial features from the X, Y coordinates by transforming them from the cartesian space to the polar space ([Reference](https://www.kaggle.com/c/sf-crime/discussion/18853))
    1. three variants of rotated Cartesian coordinates (rotated by 30, 45, 60 degree each) 
    2. Polar coordinates (i.e. the 'r' and the angle 'theta')
    3. The approach makes some intuitive sense i.e. that having such features should help in extracting some more spatial information (than relying on the current x-y alone)

In [47]:
# Normalize X and Y
print('There are %d unique longitude values, %d unique latitude values' % (train_df['X'].nunique(), 
                                                                           train_df['Y'].nunique()))

xy_scaler = StandardScaler().fit(train_df[['X', 'Y']])
train_df[['X', 'Y']] = xy_scaler.transform(train_df[['X', 'Y']])
test_df[['X', 'Y']] = xy_scaler.transform(test_df[['X', 'Y']])

There are 34309 unique longitude values, 34309 unique latitude values


In [48]:
# X-Y plane rotation and space transformation to extract more spatial information
# 2-dimensional rotation based on below functions:
# rotated x = xcos - ysin
# rotated y = xsin + ycos
# Conver from cartesian space -> polar space

cos_30 = math.cos(math.radians(30))
sin_30 = math.sin(math.radians(30))
cos_45 = math.cos(math.radians(45))
sin_45 = math.sin(math.radians(45))
cos_60 = math.cos(math.radians(60))
sin_60 = math.sin(math.radians(60))


train_df["Rot30_X"] = train_df['X'] * cos_30 - train_df['Y'] * sin_30 
train_df["Rot30_Y"] = train_df['X'] * sin_30 + train_df['Y'] * cos_30
train_df["Rot45_X"] = train_df['X'] * cos_45 - train_df['Y'] * sin_45  
train_df["Rot45_Y"] = train_df['X'] * sin_45 + train_df['Y'] * cos_45
train_df["Rot60_X"] = train_df['X'] * cos_60 - train_df['Y'] * sin_60  
train_df["Rot60_Y"] = train_df['X'] * sin_60 + train_df['Y'] * cos_60
train_df["Radius"] = np.sqrt(train_df['X'] ** 2 + train_df['Y'] ** 2)
train_df["Angle"] = np.arctan2(train_df['X'], train_df['Y'])

test_df["Rot30_X"] = test_df['X'] * cos_30 - test_df['Y'] * sin_30  
test_df["Rot30_Y"] = test_df['X'] * sin_30 + test_df['Y'] * cos_30
test_df["Rot45_X"] = test_df['X'] * cos_45 - test_df['Y'] * sin_45  
test_df["Rot45_Y"] = test_df['X'] * sin_45 + test_df['Y'] * cos_45
test_df["Rot60_X"] = test_df['X'] * cos_60 - test_df['Y'] * sin_60  
test_df["Rot60_Y"] = test_df['X'] * sin_60 + test_df['Y'] * cos_60
test_df["Radius"] = np.sqrt(test_df['X'] ** 2 + test_df['Y'] ** 2)
test_df["Angle"] = np.arctan2(test_df['X'], test_df['Y'])

In [49]:
# View the description of the numerical features again to ensure everything is right
train_df.describe()

X             Y         Minute           Hour  \
count  8.780490e+05  8.780490e+05  878049.000000  878049.000000   
mean   1.465983e-12  2.362785e-13      20.155026      13.412655   
std    1.000001e+00  1.000001e+00      18.594915       6.549573   
min   -4.834376e+00 -6.031288e+00       0.000000       0.000000   
25%   -4.028315e-01 -6.042763e-01       0.000000       9.000000   
50%    2.508318e-01  3.470638e-01      19.000000      14.000000   
75%    6.250425e-01  7.172422e-01      33.000000      19.000000   
max    3.563896e+00  4.049280e+00      59.000000      23.000000   

                 Day          Month           Year      Hour_Zone  \
count  878049.000000  878049.000000  878049.000000  878049.000000   
mean       15.570623       6.436509    2008.712046       2.212740   
std         8.783005       3.428972       3.631194       1.214321   
min         1.000000       1.000000    2003.000000       0.000000   
25%         8.000000       3.000000    2006.000000       1.000000   
50%        16.000000       6.000000    2009.000000       2.000000   
75%        23.000000       9.000000    2012.000000       3.000000   
max        31.000000      12.000000    2015.000000       4.000000   

          WeekOfYear   BusinessHour  ...        Weekend        BlockNo  \
count  878049.000000  878049.000000  ...  878049.000000  878049.000000   
mean       12.185458       0.518439  ...       0.277339       7.743576   
std         7.482940       0.499660  ...       0.447685      10.233725   
min         0.000000       0.000000  ...       0.000000       0.000000   
25%         6.000000       0.000000  ...       0.000000       0.000000   
50%        12.000000       1.000000  ...       0.000000       4.000000   
75%        19.000000       1.000000  ...       1.000000      11.000000   
max        25.000000       1.000000  ...       1.000000      84.000000   

            Rot30_X       Rot30_Y       Rot45_X       Rot45_Y       Rot60_X  \
count  8.780490e+05  8.780490e+05  8.780490e+05  8.780490e+05  8.780490e+05   
mean   1.151442e-12  9.376174e-13  8.695375e-13  1.203682e-12  5.283679e-13   
std    9.308262e-01  1.064690e+00  9.196602e-01  1.074350e+00  9.308262e-01   
min   -3.584254e+00 -6.686941e+00 -4.136693e+00 -6.845141e+00 -4.407223e+00   
25%   -4.145764e-01 -5.663592e-01 -4.910693e-01 -5.401709e-01 -5.844213e-01   
50%    6.281589e-02  2.143365e-01 -9.702803e-02  2.239870e-01 -2.373056e-01   
75%    4.804283e-01  8.071245e-01  4.085152e-01  7.867566e-01  5.062526e-01   
max    4.845782e+00  3.424154e+00  5.008162e+00  3.750343e+00  5.486811e+00   

            Rot60_Y         Radius          Angle  
count  8.780490e+05  878049.000000  878049.000000  
mean   1.387714e-12       1.219256       0.278872  
std    1.064690e+00       0.716531       1.652395  
min   -6.536856e+00       0.020040      -3.141544  
25%   -4.588610e-01       0.742410      -1.065719  
50%    3.123723e-01       1.003656       0.508769  
75%    7.511932e-01       1.639631       1.172383  
max    3.820952e+00       6.851279       3.141189  

[8 rows x 21 columns]

In [50]:
# run KMeans separately on both the training set and test set
data = [train_df, test_df]
num_clusters = 40
for dataset in data:
    coordinates = dataset.loc[:,['Y','X']]
    kmeans = KMeans(n_clusters=num_clusters, random_state=1).fit(coordinates)
    id_labels=kmeans.labels_
#     print(kmeans.cluster_centers_)
    dataset['Cluster'] = id_labels

In [51]:
train_df.head()

Dates        Category                      Descript  \
0 2015-05-13 23:53:00        WARRANTS                WARRANT ARREST   
1 2015-05-13 23:53:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
2 2015-05-13 23:33:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
3 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   
4 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   

   DayOfWeek PdDistrict      Resolution                    Address         X  \
0  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST -0.123652   
1  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST -0.123652   
2  Wednesday   NORTHERN  ARREST, BOOKED  VANNESS AV / GREENWICH ST -0.063202   
3  Wednesday   NORTHERN            NONE   1500 Block of LOMBARD ST -0.167295   
4  Wednesday       PARK            NONE  100 Block of BRODERICK ST -0.631637   

          Y  Minute  ...  BlockNo   Rot30_X   Rot30_Y   Rot45_X   Rot45_Y  \
0  0.313049      53  ...        0 -0.263610  0.209283 -0.308794  0.133924   
1  0.313049      53  ...        0 -0.263610  0.209283 -0.308794  0.133924   
2  1.381165      33  ...        0 -0.745317  1.164523 -1.021322  0.931941   
3  1.400127      30  ...       16 -0.844945  1.128898 -1.108335  0.871744   
4  0.186550      30  ...        2 -0.640288 -0.154262 -0.578545 -0.314724   

    Rot60_X   Rot60_Y    Radius     Angle  Cluster  
0 -0.332934  0.049439  0.336585 -0.376181       12  
1 -0.332934  0.049439  0.336585 -0.376181       12  
2 -1.227725  0.635848  1.382610 -0.045728        9  
3 -1.296193  0.555182  1.410087 -0.118922        9  
4 -0.477375 -0.453738  0.658609 -1.283617       12  

[5 rows x 31 columns]

## Drop Features

- We have already extracted all the necessary features from the `Address` attribute, so drop
- We don't need `Resolution` or `Descript` features since it is not included in the training data

In [52]:
# Drop Address feature from both train and test set
train_df.drop(['Address'], axis=1, inplace=True)
test_df.drop(['Address'], axis=1, inplace=True)

In [53]:
# We don't need Dates column anymore
train_df.drop(['Dates'], axis=1, inplace=True)
test_df.drop(['Dates'], axis=1, inplace=True)

In [54]:
# Drop Resolution column since test set does not have this column
train_df.drop(['Resolution'], axis=1, inplace=True)

In [55]:
# Drop Descript column since test set does not have this column
train_df.drop(['Descript'], axis=1, inplace=True)

In [56]:
# Let's quickly view the data
train_df.head()

Category  DayOfWeek PdDistrict         X         Y  Minute  Hour  \
0        WARRANTS  Wednesday   NORTHERN -0.123652  0.313049      53    23   
1  OTHER OFFENSES  Wednesday   NORTHERN -0.123652  0.313049      53    23   
2  OTHER OFFENSES  Wednesday   NORTHERN -0.063202  1.381165      33    23   
3   LARCENY/THEFT  Wednesday   NORTHERN -0.167295  1.400127      30    23   
4   LARCENY/THEFT  Wednesday       PARK -0.631637  0.186550      30    23   

   Day  Month  Year  ...  BlockNo   Rot30_X   Rot30_Y   Rot45_X   Rot45_Y  \
0   13      5  2015  ...        0 -0.263610  0.209283 -0.308794  0.133924   
1   13      5  2015  ...        0 -0.263610  0.209283 -0.308794  0.133924   
2   13      5  2015  ...        0 -0.745317  1.164523 -1.021322  0.931941   
3   13      5  2015  ...       16 -0.844945  1.128898 -1.108335  0.871744   
4   13      5  2015  ...        2 -0.640288 -0.154262 -0.578545 -0.314724   

    Rot60_X   Rot60_Y    Radius     Angle  Cluster  
0 -0.332934  0.049439  0.336585 -0.376181       12  
1 -0.332934  0.049439  0.336585 -0.376181       12  
2 -1.227725  0.635848  1.382610 -0.045728        9  
3 -1.296193  0.555182  1.410087 -0.118922        9  
4 -0.477375 -0.453738  0.658609 -1.283617       12  

[5 rows x 27 columns]

# Feature Encoding 

- Convert categorical data to numeric data

### Pd Districts

- convert Pd District categorical feature to numeric

In [57]:
pd_districts = {'SOUTHERN':0, 'MISSION':1, 'NORTHERN':2, 'CENTRAL':3, 'BAYVIEW':4, 'INGLESIDE':5, 
                'TENDERLOIN':6, 'TARAVAL':7, 'PARK':8, 'RICHMOND':9}

train_df['PdDistrict'].replace(pd_districts, inplace=True)
test_df['PdDistrict'].replace(pd_districts, inplace=True)

In [58]:
train_df.head()

Category  DayOfWeek  PdDistrict         X         Y  Minute  Hour  \
0        WARRANTS  Wednesday           2 -0.123652  0.313049      53    23   
1  OTHER OFFENSES  Wednesday           2 -0.123652  0.313049      53    23   
2  OTHER OFFENSES  Wednesday           2 -0.063202  1.381165      33    23   
3   LARCENY/THEFT  Wednesday           2 -0.167295  1.400127      30    23   
4   LARCENY/THEFT  Wednesday           8 -0.631637  0.186550      30    23   

   Day  Month  Year  ...  BlockNo   Rot30_X   Rot30_Y   Rot45_X   Rot45_Y  \
0   13      5  2015  ...        0 -0.263610  0.209283 -0.308794  0.133924   
1   13      5  2015  ...        0 -0.263610  0.209283 -0.308794  0.133924   
2   13      5  2015  ...        0 -0.745317  1.164523 -1.021322  0.931941   
3   13      5  2015  ...       16 -0.844945  1.128898 -1.108335  0.871744   
4   13      5  2015  ...        2 -0.640288 -0.154262 -0.578545 -0.314724   

    Rot60_X   Rot60_Y    Radius     Angle  Cluster  
0 -0.332934  0.049439  0.336585 -0.376181       12  
1 -0.332934  0.049439  0.336585 -0.376181       12  
2 -1.227725  0.635848  1.382610 -0.045728        9  
3 -1.296193  0.555182  1.410087 -0.118922        9  
4 -0.477375 -0.453738  0.658609 -1.283617       12  

[5 rows x 27 columns]

In [59]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 27 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Category      878049 non-null  object 
 1   DayOfWeek     878049 non-null  object 
 2   PdDistrict    878049 non-null  int64  
 3   X             878049 non-null  float64
 4   Y             878049 non-null  float64
 5   Minute        878049 non-null  int64  
 6   Hour          878049 non-null  int64  
 7   Day           878049 non-null  int64  
 8   Month         878049 non-null  int64  
 9   Year          878049 non-null  int64  
 10  Hour_Zone     878049 non-null  int64  
 11  WeekOfYear    878049 non-null  int64  
 12  Holiday       878049 non-null  bool   
 13  BusinessHour  878049 non-null  uint8  
 14  Season        878049 non-null  int64  
 15  Weekend       878049 non-null  uint8  
 16  StreetType    878049 non-null  object 
 17  BlockNo       878049 non-null  int64  
 18  Rot3

### Year

- Year is an **ordinal** variable, so let's keep that ordering and mapping
- convert Year categorical feature to numeric

In [60]:
data = [train_df, test_df]

for dataset in data:
    year_le = LabelEncoder()
    year_le.fit(dataset['Year'].unique())
    print(list(year_le.classes_))

    dataset['Year']=year_le.transform(dataset['Year']) 

[2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015]
[2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015]


In [61]:
train_df['Year'].unique()

array([12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0])

In [62]:
# So we know the mapping (important)
dict(zip(year_le.classes_, year_le.transform(year_le.classes_)))

{2003: 0,
 2004: 1,
 2005: 2,
 2006: 3,
 2007: 4,
 2008: 5,
 2009: 6,
 2010: 7,
 2011: 8,
 2012: 9,
 2013: 10,
 2014: 11,
 2015: 12}

In [63]:
train_df.head()

Category  DayOfWeek  PdDistrict         X         Y  Minute  Hour  \
0        WARRANTS  Wednesday           2 -0.123652  0.313049      53    23   
1  OTHER OFFENSES  Wednesday           2 -0.123652  0.313049      53    23   
2  OTHER OFFENSES  Wednesday           2 -0.063202  1.381165      33    23   
3   LARCENY/THEFT  Wednesday           2 -0.167295  1.400127      30    23   
4   LARCENY/THEFT  Wednesday           8 -0.631637  0.186550      30    23   

   Day  Month  Year  ...  BlockNo   Rot30_X   Rot30_Y   Rot45_X   Rot45_Y  \
0   13      5    12  ...        0 -0.263610  0.209283 -0.308794  0.133924   
1   13      5    12  ...        0 -0.263610  0.209283 -0.308794  0.133924   
2   13      5    12  ...        0 -0.745317  1.164523 -1.021322  0.931941   
3   13      5    12  ...       16 -0.844945  1.128898 -1.108335  0.871744   
4   13      5    12  ...        2 -0.640288 -0.154262 -0.578545 -0.314724   

    Rot60_X   Rot60_Y    Radius     Angle  Cluster  
0 -0.332934  0.049439  0.336585 -0.376181       12  
1 -0.332934  0.049439  0.336585 -0.376181       12  
2 -1.227725  0.635848  1.382610 -0.045728        9  
3 -1.296193  0.555182  1.410087 -0.118922        9  
4 -0.477375 -0.453738  0.658609 -1.283617       12  

[5 rows x 27 columns]

In [64]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 27 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Category      878049 non-null  object 
 1   DayOfWeek     878049 non-null  object 
 2   PdDistrict    878049 non-null  int64  
 3   X             878049 non-null  float64
 4   Y             878049 non-null  float64
 5   Minute        878049 non-null  int64  
 6   Hour          878049 non-null  int64  
 7   Day           878049 non-null  int64  
 8   Month         878049 non-null  int64  
 9   Year          878049 non-null  int64  
 10  Hour_Zone     878049 non-null  int64  
 11  WeekOfYear    878049 non-null  int64  
 12  Holiday       878049 non-null  bool   
 13  BusinessHour  878049 non-null  uint8  
 14  Season        878049 non-null  int64  
 15  Weekend       878049 non-null  uint8  
 16  StreetType    878049 non-null  object 
 17  BlockNo       878049 non-null  int64  
 18  Rot3

### DayOfWeek

- we are going to use sklearn's LabelEncoder to encode the categorical data to numeric
- Day of week is considered a categorical and nominal variable

In [65]:
data = [train_df, test_df]

for dataset in data:
    dow_le = LabelEncoder()
    dow_le.fit(dataset['DayOfWeek'].unique())
    print(list(dow_le.classes_))
    dataset['DayOfWeek']=dow_le.transform(dataset['DayOfWeek'])

['Friday', 'Monday', 'Saturday', 'Sunday', 'Thursday', 'Tuesday', 'Wednesday']
['Friday', 'Monday', 'Saturday', 'Sunday', 'Thursday', 'Tuesday', 'Wednesday']


In [66]:
train_df['DayOfWeek'].unique()

array([6, 5, 1, 3, 2, 0, 4])

In [67]:
# So we know the mapping (important)
dict(zip(dow_le.classes_, dow_le.transform(dow_le.classes_)))

{'Friday': 0,
 'Monday': 1,
 'Saturday': 2,
 'Sunday': 3,
 'Thursday': 4,
 'Tuesday': 5,
 'Wednesday': 6}

In [68]:
train_df.head()

Category  DayOfWeek  PdDistrict         X         Y  Minute  Hour  \
0        WARRANTS          6           2 -0.123652  0.313049      53    23   
1  OTHER OFFENSES          6           2 -0.123652  0.313049      53    23   
2  OTHER OFFENSES          6           2 -0.063202  1.381165      33    23   
3   LARCENY/THEFT          6           2 -0.167295  1.400127      30    23   
4   LARCENY/THEFT          6           8 -0.631637  0.186550      30    23   

   Day  Month  Year  ...  BlockNo   Rot30_X   Rot30_Y   Rot45_X   Rot45_Y  \
0   13      5    12  ...        0 -0.263610  0.209283 -0.308794  0.133924   
1   13      5    12  ...        0 -0.263610  0.209283 -0.308794  0.133924   
2   13      5    12  ...        0 -0.745317  1.164523 -1.021322  0.931941   
3   13      5    12  ...       16 -0.844945  1.128898 -1.108335  0.871744   
4   13      5    12  ...        2 -0.640288 -0.154262 -0.578545 -0.314724   

    Rot60_X   Rot60_Y    Radius     Angle  Cluster  
0 -0.332934  0.049439  0.336585 -0.376181       12  
1 -0.332934  0.049439  0.336585 -0.376181       12  
2 -1.227725  0.635848  1.382610 -0.045728        9  
3 -1.296193  0.555182  1.410087 -0.118922        9  
4 -0.477375 -0.453738  0.658609 -1.283617       12  

[5 rows x 27 columns]

In [69]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 27 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Category      878049 non-null  object 
 1   DayOfWeek     878049 non-null  int64  
 2   PdDistrict    878049 non-null  int64  
 3   X             878049 non-null  float64
 4   Y             878049 non-null  float64
 5   Minute        878049 non-null  int64  
 6   Hour          878049 non-null  int64  
 7   Day           878049 non-null  int64  
 8   Month         878049 non-null  int64  
 9   Year          878049 non-null  int64  
 10  Hour_Zone     878049 non-null  int64  
 11  WeekOfYear    878049 non-null  int64  
 12  Holiday       878049 non-null  bool   
 13  BusinessHour  878049 non-null  uint8  
 14  Season        878049 non-null  int64  
 15  Weekend       878049 non-null  uint8  
 16  StreetType    878049 non-null  object 
 17  BlockNo       878049 non-null  int64  
 18  Rot3

### Street Type

- we are going to use sklearn's LabelEncoder to encode the categorical data to numeric

In [70]:
data = [train_df, test_df]

for dataset in data:
    st_le = LabelEncoder()
    st_le.fit(dataset['StreetType'].unique())
    print(list(st_le.classes_))
    dataset['StreetType']=st_le.transform(dataset['StreetType'])

['AL', 'AV', 'BL', 'CR', 'CT', 'DR', 'EL CAMINO DEL MAR', 'HY', 'I-80', 'INT', 'LN', 'OTHER', 'PL', 'PZ', 'RD', 'RW', 'ST', 'TR', 'WAY', 'WK', 'WY']
['AL', 'AV', 'BL', 'CR', 'CT', 'DR', 'EL CAMINO DEL MAR', 'HY', 'I-80', 'INT', 'LN', 'OTHER', 'PL', 'PZ', 'RD', 'RW', 'ST', 'TR', 'WAY', 'WK', 'WY']


In [71]:
train_df['StreetType'].unique()

array([ 9, 16,  1, 20,  4,  5, 17,  7, 14, 13,  2, 12, 10, 18,  3,  0, 11,
       19, 15,  6,  8])

In [72]:
train_df.head()

Category  DayOfWeek  PdDistrict         X         Y  Minute  Hour  \
0        WARRANTS          6           2 -0.123652  0.313049      53    23   
1  OTHER OFFENSES          6           2 -0.123652  0.313049      53    23   
2  OTHER OFFENSES          6           2 -0.063202  1.381165      33    23   
3   LARCENY/THEFT          6           2 -0.167295  1.400127      30    23   
4   LARCENY/THEFT          6           8 -0.631637  0.186550      30    23   

   Day  Month  Year  ...  BlockNo   Rot30_X   Rot30_Y   Rot45_X   Rot45_Y  \
0   13      5    12  ...        0 -0.263610  0.209283 -0.308794  0.133924   
1   13      5    12  ...        0 -0.263610  0.209283 -0.308794  0.133924   
2   13      5    12  ...        0 -0.745317  1.164523 -1.021322  0.931941   
3   13      5    12  ...       16 -0.844945  1.128898 -1.108335  0.871744   
4   13      5    12  ...        2 -0.640288 -0.154262 -0.578545 -0.314724   

    Rot60_X   Rot60_Y    Radius     Angle  Cluster  
0 -0.332934  0.049439  0.336585 -0.376181       12  
1 -0.332934  0.049439  0.336585 -0.376181       12  
2 -1.227725  0.635848  1.382610 -0.045728        9  
3 -1.296193  0.555182  1.410087 -0.118922        9  
4 -0.477375 -0.453738  0.658609 -1.283617       12  

[5 rows x 27 columns]

In [73]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 27 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Category      878049 non-null  object 
 1   DayOfWeek     878049 non-null  int64  
 2   PdDistrict    878049 non-null  int64  
 3   X             878049 non-null  float64
 4   Y             878049 non-null  float64
 5   Minute        878049 non-null  int64  
 6   Hour          878049 non-null  int64  
 7   Day           878049 non-null  int64  
 8   Month         878049 non-null  int64  
 9   Year          878049 non-null  int64  
 10  Hour_Zone     878049 non-null  int64  
 11  WeekOfYear    878049 non-null  int64  
 12  Holiday       878049 non-null  bool   
 13  BusinessHour  878049 non-null  uint8  
 14  Season        878049 non-null  int64  
 15  Weekend       878049 non-null  uint8  
 16  StreetType    878049 non-null  int64  
 17  BlockNo       878049 non-null  int64  
 18  Rot3

### Holiday

- Encode the binary feature

In [74]:
# Encode to 0 and 1

train_df['Holiday'].replace(False, 0, inplace=True)
train_df['Holiday'].replace(True, 1, inplace=True)
test_df['Holiday'].replace(False, 0, inplace=True)
test_df['Holiday'].replace(True, 1, inplace=True)

train_df['Holiday'] = train_df['Holiday'].astype('uint8')
train_df['Holiday'] = train_df['Holiday'].astype('uint8')

In [75]:
train_df[train_df['Holiday'] == 1].head()

Category  DayOfWeek  PdDistrict         X         Y  Minute  \
18357   DRUG/NARCOTIC          1           0  0.765578  0.347064      48   
18358  OTHER OFFENSES          1           0  0.765578  0.347064      48   
18359    NON-CRIMINAL          1           5  0.436364 -2.363341      30   
18360       VANDALISM          1           3  0.281633  1.685193      30   
18361   LARCENY/THEFT          1           2 -0.250919  0.118598      30   

       Hour  Day  Month  Year  ...  BlockNo   Rot30_X   Rot30_Y   Rot45_X  \
18357    23   16      2    12  ...        9  0.489478  0.683355  0.295934   
18358    23   16      2    12  ...        9  0.489478  0.683355  0.295934   
18359    23   16      2    12  ...        3  1.559573 -1.828532  1.979690   
18360    23   16      2    12  ...       28 -0.598695  1.600236 -0.992467   
18361    23   16      2    12  ...        1 -0.276602 -0.022750 -0.261289   

        Rot45_Y   Rot60_X   Rot60_Y    Radius     Angle  Cluster  
18357  0.786757  0.082223  0.836542  0.840573  1.145172       31  
18358  0.786757  0.082223  0.836542  0.840573  1.145172       31  
18359 -1.362579  2.264895 -0.803768  2.403288  2.959010       15  
18360  1.390756 -1.318603  1.086497  1.708564  0.165592        6  
18361 -0.093565 -0.228169 -0.158003  0.277536 -1.129263       12  

[5 rows x 27 columns]

In [76]:
test_df[test_df['Holiday'] == 1].head()

Id  DayOfWeek  PdDistrict         X         Y  Minute  Hour  Day  \
27957  27957          4           3  0.499745  1.574970      55    23    1   
27958  27958          4           6  0.192217  0.750317      43    23    1   
27959  27959          4           1 -0.121694 -0.265227      30    23    1   
27960  27960          4           7 -2.341083 -1.309824      15    23    1   
27961  27961          4           7 -2.341083 -1.309824      15    23    1   

       Month  Year  ...  BlockNo   Rot30_X   Rot30_Y   Rot45_X   Rot45_Y  \
27957      1    12  ...        0 -0.354693  1.613837 -0.760299  1.467045   
27958      1    12  ...        0 -0.208694  0.745902 -0.394636  0.666472   
27959      1    12  ...        6  0.027224 -0.290540  0.101493 -0.273594   
27960      1    12  ...        1 -1.372526 -2.304882 -0.729211 -2.581581   
27961      1    12  ...        1 -1.372526 -2.304882 -0.729211 -2.581581   

        Rot60_X   Rot60_Y    Radius     Angle  Cluster  
27957 -1.114092  1.220277  1.652355  0.307256        9  
27958 -0.553685  0.541623  0.774546  0.250788        1  
27959  0.168846 -0.238003  0.291813 -2.711421        3  
27960 -0.036201 -2.682350  2.682594 -2.080900       13  
27961 -0.036201 -2.682350  2.682594 -2.080900       13  

[5 rows x 27 columns]

### Category

- we are going to use sklearn's LabelEncoder to encode the categorical data to numeric

In [77]:
data = [train_df]

for dataset in data:
    cat_le = LabelEncoder()
    cat_le.fit(dataset['Category'].unique())
    print(list(cat_le.classes_))
    dataset['Category']=cat_le.transform(dataset['Category'])

['ARSON', 'ASSAULT', 'BAD CHECKS', 'BRIBERY', 'BURGLARY', 'DISORDERLY CONDUCT', 'DRIVING UNDER THE INFLUENCE', 'DRUG/NARCOTIC', 'DRUNKENNESS', 'EMBEZZLEMENT', 'EXTORTION', 'FAMILY OFFENSES', 'FORGERY/COUNTERFEITING', 'FRAUD', 'GAMBLING', 'KIDNAPPING', 'LARCENY/THEFT', 'LIQUOR LAWS', 'LOITERING', 'MISSING PERSON', 'NON-CRIMINAL', 'OTHER OFFENSES', 'PORNOGRAPHY/OBSCENE MAT', 'PROSTITUTION', 'RECOVERED VEHICLE', 'ROBBERY', 'RUNAWAY', 'SECONDARY CODES', 'SEX OFFENSES FORCIBLE', 'SEX OFFENSES NON FORCIBLE', 'STOLEN PROPERTY', 'SUICIDE', 'SUSPICIOUS OCC', 'TREA', 'TRESPASS', 'VANDALISM', 'VEHICLE THEFT', 'WARRANTS', 'WEAPON LAWS']


In [78]:
len(train_df['Category'].unique())

39

In [79]:
# So we know the mapping (important)
dict(zip(cat_le.classes_, cat_le.transform(cat_le.classes_)))

{'ARSON': 0,
 'ASSAULT': 1,
 'BAD CHECKS': 2,
 'BRIBERY': 3,
 'BURGLARY': 4,
 'DISORDERLY CONDUCT': 5,
 'DRIVING UNDER THE INFLUENCE': 6,
 'DRUG/NARCOTIC': 7,
 'DRUNKENNESS': 8,
 'EMBEZZLEMENT': 9,
 'EXTORTION': 10,
 'FAMILY OFFENSES': 11,
 'FORGERY/COUNTERFEITING': 12,
 'FRAUD': 13,
 'GAMBLING': 14,
 'KIDNAPPING': 15,
 'LARCENY/THEFT': 16,
 'LIQUOR LAWS': 17,
 'LOITERING': 18,
 'MISSING PERSON': 19,
 'NON-CRIMINAL': 20,
 'OTHER OFFENSES': 21,
 'PORNOGRAPHY/OBSCENE MAT': 22,
 'PROSTITUTION': 23,
 'RECOVERED VEHICLE': 24,
 'ROBBERY': 25,
 'RUNAWAY': 26,
 'SECONDARY CODES': 27,
 'SEX OFFENSES FORCIBLE': 28,
 'SEX OFFENSES NON FORCIBLE': 29,
 'STOLEN PROPERTY': 30,
 'SUICIDE': 31,
 'SUSPICIOUS OCC': 32,
 'TREA': 33,
 'TRESPASS': 34,
 'VANDALISM': 35,
 'VEHICLE THEFT': 36,
 'WARRANTS': 37,
 'WEAPON LAWS': 38}

In [80]:
train_df.head()

Category  DayOfWeek  PdDistrict         X         Y  Minute  Hour  Day  \
0        37          6           2 -0.123652  0.313049      53    23   13   
1        21          6           2 -0.123652  0.313049      53    23   13   
2        21          6           2 -0.063202  1.381165      33    23   13   
3        16          6           2 -0.167295  1.400127      30    23   13   
4        16          6           8 -0.631637  0.186550      30    23   13   

   Month  Year  ...  BlockNo   Rot30_X   Rot30_Y   Rot45_X   Rot45_Y  \
0      5    12  ...        0 -0.263610  0.209283 -0.308794  0.133924   
1      5    12  ...        0 -0.263610  0.209283 -0.308794  0.133924   
2      5    12  ...        0 -0.745317  1.164523 -1.021322  0.931941   
3      5    12  ...       16 -0.844945  1.128898 -1.108335  0.871744   
4      5    12  ...        2 -0.640288 -0.154262 -0.578545 -0.314724   

    Rot60_X   Rot60_Y    Radius     Angle  Cluster  
0 -0.332934  0.049439  0.336585 -0.376181       12  
1 -0.332934  0.049439  0.336585 -0.376181       12  
2 -1.227725  0.635848  1.382610 -0.045728        9  
3 -1.296193  0.555182  1.410087 -0.118922        9  
4 -0.477375 -0.453738  0.658609 -1.283617       12  

[5 rows x 27 columns]

In [81]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 27 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Category      878049 non-null  int64  
 1   DayOfWeek     878049 non-null  int64  
 2   PdDistrict    878049 non-null  int64  
 3   X             878049 non-null  float64
 4   Y             878049 non-null  float64
 5   Minute        878049 non-null  int64  
 6   Hour          878049 non-null  int64  
 7   Day           878049 non-null  int64  
 8   Month         878049 non-null  int64  
 9   Year          878049 non-null  int64  
 10  Hour_Zone     878049 non-null  int64  
 11  WeekOfYear    878049 non-null  int64  
 12  Holiday       878049 non-null  uint8  
 13  BusinessHour  878049 non-null  uint8  
 14  Season        878049 non-null  int64  
 15  Weekend       878049 non-null  uint8  
 16  StreetType    878049 non-null  int64  
 17  BlockNo       878049 non-null  int64  
 18  Rot3

## View Information about Data

- One last check before training

In [82]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 27 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Category      878049 non-null  int64  
 1   DayOfWeek     878049 non-null  int64  
 2   PdDistrict    878049 non-null  int64  
 3   X             878049 non-null  float64
 4   Y             878049 non-null  float64
 5   Minute        878049 non-null  int64  
 6   Hour          878049 non-null  int64  
 7   Day           878049 non-null  int64  
 8   Month         878049 non-null  int64  
 9   Year          878049 non-null  int64  
 10  Hour_Zone     878049 non-null  int64  
 11  WeekOfYear    878049 non-null  int64  
 12  Holiday       878049 non-null  uint8  
 13  BusinessHour  878049 non-null  uint8  
 14  Season        878049 non-null  int64  
 15  Weekend       878049 non-null  uint8  
 16  StreetType    878049 non-null  int64  
 17  BlockNo       878049 non-null  int64  
 18  Rot3

In [83]:
# # Convert all to 32 bit integers so less memory and will train faster (no loss in data since our integers dont reach)
columns_to_convert = ['DayOfWeek', 'PdDistrict', 'Minute', 'Hour', 'Day', 'Month', 'Year', 
                      'Hour_Zone', 'WeekOfYear', 'Season', 'StreetType', 'BlockNo', 'Cluster']
train_df[columns_to_convert] = train_df[columns_to_convert].astype('int16')
test_df[columns_to_convert] = test_df[columns_to_convert].astype('int16')

train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878049 entries, 0 to 878048
Data columns (total 27 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Category      878049 non-null  int64  
 1   DayOfWeek     878049 non-null  int16  
 2   PdDistrict    878049 non-null  int16  
 3   X             878049 non-null  float64
 4   Y             878049 non-null  float64
 5   Minute        878049 non-null  int16  
 6   Hour          878049 non-null  int16  
 7   Day           878049 non-null  int16  
 8   Month         878049 non-null  int16  
 9   Year          878049 non-null  int16  
 10  Hour_Zone     878049 non-null  int16  
 11  WeekOfYear    878049 non-null  int16  
 12  Holiday       878049 non-null  uint8  
 13  BusinessHour  878049 non-null  uint8  
 14  Season        878049 non-null  int16  
 15  Weekend       878049 non-null  uint8  
 16  StreetType    878049 non-null  int16  
 17  BlockNo       878049 non-null  int16  
 18  Rot3

# Building Machine Learning Models

- Baseline Models
    - Let's train a couple models on a stratified sample of the training data
    - Evaluate on a hold out set to get baseline results for each model to determine what model to use
    - Models:
        - Stochastic Gradient Descent (with elastic net regularization)
        - Gaussian Naive Bayes
        - K Nearest Neighbors
        - Logistic Regression (with L1 regularization)
        - Random Forest
        - XGBoost
    - Almost all the default scikit-learn ML algorithm hyperparameters exhibit bad performance
        - Researched online & read literature to determine some more ideal default hyperparameters
            - [Reference](https://arxiv.org/abs/1708.05070)
- Couple things to note:
    - **Decision tree models** including Ensemble methods (Random Forest & XGBoost) can handle categorical variables without one-hot encoding them. 
    - **Linear models** (SGD & Logistic Regression) cannot handle categorical features & need features to be OHE before training
    - Always OneHotEncode before you split data up to training/dev/test so that all features & classes will be represented

In [84]:
# Set training data (drop labels) and training labels
X_train = train_df.drop("Category", axis=1).copy()
Y_train = train_df["Category"].copy()

# Set testing data (drop Id)
X_test = test_df.drop("Id", axis=1).copy()

In [85]:
# Use these for ML algorithms that can handle categorical data without OHE
mini_train_data, mini_dev_data, mini_train_labels, mini_dev_labels = train_test_split(X_train, 
                                                                                      Y_train,
                                                                                      stratify=Y_train,
                                                                                      test_size=0.5,
                                                                                      random_state=1)

# Hyperparameter Tuning

- Hyperparameter tuning involves defining an objective function (log loss), and using cross-validation to measure the hyperparameter quality. 
    - We want the hyperparameters that give the highest generalization performance.
- Three approaches: Grid Search (`GridSearchCV`), Random Search (`RandomSearchCV`), and Bayes Optimization (`BayesSearchCV`)
- Realized `GridSearchCV` took way too long and was impractical, and `RandomSearchCV` was too random.
    - Grid and random search are completely uninformed by past evaluations, and as a result, often spend a significant amount of time evaluating “bad” hyperparameters.
- Then, I did more research on more efficient & smarter hyperparameter tuning techniques and found Bayeisan Optimization (`BayesSearchCV`)
- **Bayesian Optimization Overview**
    - Build a probabilistic model of the objective function & use it to select promising hyperparameters to evaluate in the true objective function
        - The model used for approximating the objective function is called *surrogate model*. 
            - E.g. Gaussian Processes 
    - Keeps track of past evaluation results, which is used to form a probabilistic model mapping hyperparameters to a probability of a score on the objective function
    - Instead of optimizing an expensive objective function, we optimize on a cheap proxy function instead.
        - *Acquisition function* that directs sampling to areas where an improvement over the current best observation is likely.
            - E.g. maximum probability of improvement (MPI), expected improvement (EI) and upper confidence bound (UCB)
- **K-Folds Cross Validation**
    - Use cross validation to measure the true generalization performance of a model 
    - This is integrated with the hyperparameter tuning techniques (`GridSearchCV`, `RandomSearchCV`, `BayesSearchCV`)

## XGBoost (Boosting)

- Basic Overview:
    - Another ensemble method that uses Boosting instead of Bagging (Random Forests)
    - In **Boosting**, the trees are built sequentially such that each subsequent tree aims to reduce the errors of the previous tree.
    - Each tree learns from its predecessors and updates the residual errors. 
    - Each base learner is weak (high bias) and contributes some vital information for prediction, enabling the boosting technique to produce a strong learner by effectively combining these weak learners.
    - The final strong learner brings down both the **bias** and the **variance**.
    - In contrast to bagging techniques like Random Forest, in which trees are grown to their maximum extent, boosting makes use of trees with fewer splits
        -  Such small trees, which are not very deep, are **highly interpretable**. 
- Basic Steps:
    1. Initial model `F0` to predict target variable `y`. Used to also calculate residual (`y - F0`)
    2. A new model `h1` is used to fit to the residuals from the previous step
    3. Now, `F0` and `h1` are combined to give `F1`, which is the boosted version of `F0`. 
        - The MSE or whatever cost function you use (Log loss, MAE) of `F1` will be lower than `F0`.
    4. Iterate the above steps to create new models based off the previous models.
    
### Prevent Overfitting:
- Large number of trees will cause overfitting (unlike Random Forests)


# Train model with optimal hyperparameters & all features

- Initially, I started with a Random Forest, but decided to use XGBoost
- We first train the model (with all the features) using the optimal hyperparameters that were found through `BayesSearchCV`
- Then, I use the model to predict the probabilities of test set with all the features
    - I'll save these predictions later to compare them with another model I will train with certain features removed

In [86]:
# It seems running time scales quadratically with the number of classes
xgb = XGBClassifier(
    n_estimators=86, 
    objective="multi:softprob", 
    learning_rate=0.1858621466840661,
    colsample_bylevel=1.0,
    colsample_bytree=1.0,
    gamma=0.49999999999999994,
    max_delta_step=0,
    max_depth=50,
    min_child_weight=5,
    reg_alpha=1.0,
    reg_lambda=60.121460571845695,
    scale_pos_weight=1e-06,
    subsample=1.0,
    random_state=1, 
    tree_method="gpu_hist",
    gpu_id=0,
    n_jobs=4,
    verbosity=0)


xgb.fit(X_train, Y_train)

Y_test_pred = xgb.predict_proba(X_test)

# Feature Importance

- Measured by mean decrease in Gini information
- This is a form of feature selection that ensemble methods (Random Forest, XGBoost) can use to prevent overfitting
    - I drop the features that seem unimportant & with less than a 1% contribution

In [87]:
importances = pd.DataFrame({'feature': X_train.columns,
                            'importance': np.round(xgb.feature_importances_, 5)})
importances = importances.sort_values('importance',ascending=False).set_index('feature')

In [88]:
importances

importance
feature                 
Minute           0.10050
Cluster          0.08395
BusinessHour     0.08303
BlockNo          0.06674
Rot45_Y          0.04846
Hour_Zone        0.04567
PdDistrict       0.04103
Rot60_X          0.03892
Rot30_Y          0.03889
Radius           0.03748
Rot30_X          0.03482
Year             0.03366
Hour             0.03257
X                0.03225
Y                0.03174
Rot45_X          0.03113
Rot60_Y          0.03056
Angle            0.03002
Weekend          0.02522
StreetType       0.02501
Season           0.01875
WeekOfYear       0.01866
DayOfWeek        0.01843
Day              0.01783
Month            0.01774
Holiday          0.01695

# Model Evaluation

- Evaluate final model based on K-Fold cross validation
- Average all K iterations to give the true estimate of the final model's performance

In [97]:
scores = cross_val_score(xgb, X_train, Y_train, 
                         cv=StratifiedKFold(n_splits=3), 
                         scoring = "neg_log_loss", n_jobs=2)

In [98]:
print("Scores:", scores)
print("Mean:", scores.mean())
print("Standard Deviation:", scores.std())

Scores: [-5.54253493 -4.80646293 -8.04966467]
Mean: -6.132887511275459
Standard Deviation: 1.3882786750052039


# Kaggle Submission

- Reformat and turn in predictions and results from our model

In [91]:
Y_test_pred.shape

(884262, 39)

In [92]:
sample_submission = pd.read_csv('data/sampleSubmission.csv')

In [93]:
sample_submission.shape

(884262, 40)

In [94]:
sample_submission.iloc[:, 1:] = pd.DataFrame(Y_test_pred, columns=sample_submission.columns[1:])

In [95]:
sample_submission.head(10)

Id     ARSON   ASSAULT  BAD CHECKS   BRIBERY  BURGLARY  DISORDERLY CONDUCT  \
0   0  0.003804  0.098074    0.000142  0.000878  0.030028            0.002153   
1   1  0.001043  0.036833    0.000066  0.000511  0.001067            0.001752   
2   2  0.003743  0.113474    0.000164  0.000384  0.070573            0.001593   
3   3  0.002119  0.202444    0.000183  0.002068  0.038323            0.004902   
4   4  0.002119  0.202444    0.000183  0.002068  0.038323            0.004902   
5   5  0.002195  0.110555    0.000084  0.000425  0.001491            0.001233   
6   6  0.002318  0.032387    0.000078  0.000746  0.075137            0.001398   
7   7  0.002543  0.064465    0.000054  0.000474  0.073691            0.000763   
8   8  0.003204  0.173838    0.000144  0.002029  0.034154            0.005515   
9   9  0.000861  0.076738    0.000088  0.000162  0.004658            0.001435   

   DRIVING UNDER THE INFLUENCE  DRUG/NARCOTIC  DRUNKENNESS  ...  \
0                     0.010237       0.019319     0.003591  ...   
1                     0.012197       0.072491     0.001670  ...   
2                     0.001825       0.017748     0.010431  ...   
3                     0.005343       0.028699     0.012530  ...   
4                     0.005343       0.028699     0.012530  ...   
5                     0.012451       0.012889     0.005738  ...   
6                     0.002189       0.005160     0.001302  ...   
7                     0.000891       0.004314     0.001216  ...   
8                     0.006582       0.039087     0.019293  ...   
9                     0.005165       0.014407     0.007984  ...   

   SEX OFFENSES NON FORCIBLE  STOLEN PROPERTY   SUICIDE  SUSPICIOUS OCC  \
0                   0.000180         0.004221  0.000925        0.074013   
1                   0.000133         0.003880  0.000089        0.037469   
2                   0.000104         0.007436  0.000879        0.058209   
3                   0.000256         0.008732  0.000706        0.084481   
4                   0.000256         0.008732  0.000706        0.084481   
5                   0.000170         0.004246  0.000141        0.027605   
6                   0.000153         0.003947  0.000575        0.024767   
7                   0.000066         0.001968  0.000189        0.017973   
8                   0.000160         0.009857  0.000462        0.049637   
9                   0.000092         0.009988  0.000137        0.019002   

       TREA  TRESPASS  VANDALISM  VEHICLE THEFT  WARRANTS  WEAPON LAWS  
0  0.000039  0.005711   0.042537       0.285357  0.029328     0.035950  
1  0.000026  0.000524   0.009592       0.021212  0.110628     0.016044  
2  0.000034  0.008407   0.055654       0.020026  0.024829     0.007298  
3  0.000042  0.004065   0.065912       0.092033  0.055587     0.019618  
4  0.000042  0.004065   0.065912       0.092033  0.055587     0.019618  
5  0.000033  0.000558   0.047966       0.048101  0.064682     0.019471  
6  0.000027  0.003088   0.083210       0.357542  0.006450     0.005805  
7  0.000019  0.002720   0.040606       0.461149  0.003812     0.003978  
8  0.000041  0.009115   0.057322       0.010319  0.073970     0.015292  
9  0.000032  0.000860   0.040035       0.019747  0.023711     0.006540  

[10 rows x 40 columns]

In [96]:
sample_submission.to_csv('data/submission_baseline.csv', index=False)

KeyboardInterrupt: 

# Summary

- After lots of tuning, I finally achieved a kaggle evaluation score (multiclass log loss) of **2.25674**, which would ideally rank at **#136** (out of 2,335 teams) or at the **top 6%** or **94th percentile** on the public leaderboard
    - Since this is an old kaggle competition, this would most likely be a lower rank, but I still felt proud to achieve this score
    - It is possible that I could run more experiments and tune the hyperparameters to achieve an even better score & ranking 
    - This was more of a learning experience for me & to get my feet wet with Data Science projects & Kaggle competitions
    - In an effort to learn, I refrained from looking up old Kaggle kernels & other resources that completed this specific Kaggle competition.
    - I coded most of this myself to learn the data science libraries, but did use resources such as other Kaggle competition kernels and research papers to get a better idea of how to think about the data. Google is awesome.
- Below, I show images of my two highest scoring submissions on Kaggle

# Conclusion

This project has taught me a lot about data science and has given me hands-on experience with working with data and completing an end-to-end data science project. I've had a lot of fun visualizing, analyzing, and experimenting with the data to gain more insight. This is just the beginning of my journey into data science, and I am very excited to see what the future holds in terms of new and interesting data science problems and datasets.

- **What I learned**:
    - There are more efficient ways to label or integer encode features
        - Will use sklearn's LabelEncoder, OneHotEncoder, & MultiLabelBinarizer next time
    - Instead of just blindly training models, research more about ways to optimize the hyperparameters efficiently
        - Spent too many AWS EC2 hours with `GridSearchCV`, when I should have used *Bayesian Optimization* for efficient hyperparameter tuning
        - Do more research on the domain of the problem, certain core ML algorithms, and data processing techniques
- **What's next?**
    - AutoML with `tpot` or `auto-sklearn`
        - automate the hyperparameter tuning and model selection with AutoML packages
    - Problem Redirection (Classification ---> Regression)
        - Instead of predicting category of crime, predict X & Y coordinates (longitude & latitude) continuous values given same spatial and temporal features as well as category of crime
        - **Use case:** Dynamically concentrate police on certain serious categories of crime to prevent crimes from happening beforehand
    - Rewrite all code in the jupyter notebook to .py files
        - Modularize each of the steps with functions and/or classes
        - Useful because I can run the .py file on AWS EC2 without having to host it on jupyter notebook locally
            - Meaning I can peacefully shut down my laptop and let script run in the cloud overnight
